### PR TITLE
hledger 1.20

### PR DESCRIPTION
**Pull request by simonmichael**
Update to hledger 1.20. Also,
- use hledger-web's built-in tests
- build with a newer stackage snapshot.

### CODE REVIEW

File: Formula/hledger.rb [hunk 1]
```diff
@@ -45,18 +45,30 @@ def install
     (buildpath/"../hledger-web").install resource("hledger-web")
     cd ".." do
       system "stack", "update"
-      system "stack", "init", "--resolver=lts-16.12"
+      File.open("stack.yaml", "w") do |f|
+        f.write <<~EOS
```
Comment (carlocab) [inline Formula/hledger.rb:49]:
```suggestion
      (buildpath/"stack.yaml").write <<~EOS
```
This should also work, no?

Comment (simonmichael) [reply Formula/hledger.rb:49]: Thanks. How would I tell it ./stack.yaml ?

Comment (carlocab) [reply Formula/hledger.rb:49]: Edited my suggestion. Try that. But no longer sure; not sure why these things are happening in the parent of `buildpath`.

Comment (simonmichael) [reply Formula/hledger.rb:49]: stack expects to see all four haskell packages - hledger-1.20 (buildpath) and hledger-lib, hledger-ui, hledger-web - as subdirectories. For now at least.

I have a mystery problem preventing me from retesting; I'll come back to this.

**Comment by carlocab** [comment]
Ok, that seems to have broken this formula and I don't really understand why. Please change it back. Sorry!

**Comment by chenrui333** [comment]
should be good now for the PR build.

**Comment by chenrui333** [comment]
only failed on 11.0 now.

**Comment by carlocab** [comment]
```
--  While building package hledger-ui-1.20 using:
      /private/tmp/hledger-20201210-36773-j90789/hledger-1.20/.brew_home/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_3.0.1.0_ghc-8.8.4 --builddir=.stack-work/dist/x86_64-osx/Cabal-3.0.1.0 build exe:hledger-ui --ghc-options ""
14812
    Process exited with code: ExitFailure 1
```

**Comment by simonmichael** [comment]
This is the first time I've built Haskell on macos 11, and the problem seems to be https://gitlab.haskell.org/ghc/ghc/-/issues/18446, affecting all current GHC releases. How are other haskell packages in homebrew dealing with this I wonder ? Maybe they don't require system frameworks ?

**Comment by simonmichael** [comment]
It seems like the options for updating (affected?) haskell apps in homebrew right now are: 
- wait for the GHC 8.8.5/8.10.3 release; or 
- try [this workaround](https://github.com/yairchu/macos11-haskell-workaround), involving custom C code which I guess would have to be included in the checksummed tarball.

**Comment by carlocab** [comment]
Does the workaround work for you locally?

One way to use it when installing formulae would be to declare it in a `resource` block, though you'll want to choose a specific commit has to refer to in that.

Also, if upstream already have fixes lined up for the next release, it might be better to just use their patches with existing formulae.

Pinging @SMillerDev, who'll know better about whether a workaround is the right thing to do for Homebrew.

**Comment by simonmichael** [comment]
I don't have big sur locally. Thanks for the info, I'll wait to hear from SMiller, but I am inclined to hope that a GHC update will be released and become available in homebrew soon enough.

**Comment by SMillerDev** [comment]
Whatever is an official patch from upstream we'd want to use.

**Comment by carlocab** [comment]
>the problem seems to be https://gitlab.haskell.org/ghc/ghc/-/issues/18446, affecting all current GHC releases

This seems to be affecting a number of other haskell-based builds too. But I suppose that shouldn't be surprising. See https://github.com/Homebrew/homebrew-core/pull/65759, though I am essentially just repeating @simonmichael's comments there.

**Comment by carlocab** [comment]
This should hopefully be unblocked by https://github.com/Homebrew/homebrew-core/pull/67284

**Comment by carlocab** [comment]
Picked up your commits, squashed them and included them here: https://github.com/Homebrew/homebrew-core/pull/67284

Hopefully, it fixes some build failures...

**Comment by carlocab** [comment]
Actually, you can add whatever fixes to `hledger` so that it builds on ghc 8.10.3 here too. I'll be able to incorporate them into my ghc PR.

**Comment by simonmichael** [comment]
Updated to use homebrew's latest ghc; CI is expected to fail for this push, but hopefully will succeed when it's combined with #67284.

**Comment by carlocab** [comment]
Thanks. I've added your commit to the ghc PR.

**Comment by simonmichael** [comment]
Switched back to cabal (temporarily ?) since stack + ghc 8.10.3 isn't working just now.

**Comment by carlocab** [comment]
Can you try replacing `install_cabal_package` to the appropriate calls to `system "cabal"`, etc here?

File: Formula/hledger.rb [hunk 2]
```diff
@@ -45,18 +45,30 @@ def install
...
+        compiler: ghc-8.10.3
+        compiler-check: newer-minor
+        packages:
+        - hledger-1.20.1
+        - hledger-lib
+        - hledger-ui
+        - hledger-web
+        extra-deps:
+        - pretty-simple-4.0.0.0
+        - prettyprinter-1.7.0
```
Comment (carlocab) [inline Formula/hledger.rb:59]: Could you leave a brief comment before this that explains how these values are supposed to be updated? That way the next person who updates this formula can easily make the needed changes. Also:
```suggestion
        resolver: lts-16.27
        compiler: ghc-#{Formula["ghc"].version}
        compiler-check: newer-minor
        packages:
        - hledger-#{version}
        - hledger-lib
        - hledger-ui
        - hledger-web
        extra-deps:
        - pretty-simple-4.0.0.0
        - prettyprinter-1.7.0
```

File: Formula/hledger.rb [hunk 3]
```diff
@@ -45,18 +45,30 @@ def install
...
+      EOS
+      system "stack", "install", "--system-ghc", "--no-install-ghc", "--skip-ghc-check", "--local-bin-path=#{bin}"
 
-      man1.install "hledger-1.19.1/hledger.1"
-      man1.install "hledger-ui/hledger-ui.1"
-      man1.install "hledger-web/hledger-web.1"
+      man1.install "hledger-1.20.1/hledger.1"
```
Comment (carlocab) [inline Formula/hledger.rb:63]:
```suggestion
      man1.install "hledger-#{version}/hledger.1"
```

File: Formula/hledger.rb [hunk 4]
```diff
@@ -45,18 +45,30 @@ def install
...
       man5.install "hledger-lib/hledger_csv.5"
       man5.install "hledger-lib/hledger_journal.5"
       man5.install "hledger-lib/hledger_timeclock.5"
       man5.install "hledger-lib/hledger_timedot.5"
+      man1.install "hledger-ui/hledger-ui.1"
+      man1.install "hledger-web/hledger-web.1"
 
-      info.install "hledger-1.19.1/hledger.info"
+      info.install "hledger-1.20.1/hledger.info"
```
Comment (carlocab) [inline Formula/hledger.rb:71]:
```suggestion
      info.install "hledger-#{version}/hledger.info"
```

File: Formula/hledger.rb [hunk 5]
```diff
@@ -20,23 +20,23 @@ class Hledger < Formula
...
-    url "https://hackage.haskell.org/package/hledger-ui-1.19.1/hledger-ui-1.19.1.tar.gz"
-    sha256 "d4154c33712b003dc15d795c92da59158d2ca5e02660234f731d2794a5403f9e"
+    url "https://hackage.haskell.org/package/hledger-ui-1.20.1/hledger-ui-1.20.1.tar.gz"
+    sha256 "cc1b0b307e89d525b6393a5c07733f531cf13e69eb9235e3aef2c7fdbf0fa737"
   end
   resource "hledger-web" do
-    url "https://hackage.haskell.org/package/hledger-web-1.19.1/hledger-web-1.19.1.tar.gz"
-    sha256 "6085cb69bdc3808f929cc6ee621c0d3ffc773debe42bc1aaf0c7c1fe1a988a0f"
+    url "https://hackage.haskell.org/package/hledger-web-1.20.1/hledger-web-1.20.1.tar.gz"
+    sha256 "14c93f228e28fc63122db7b2b6af70fb15a581f612165b6a9ee46ee6f6789b68"
```
Comment (carlocab) [inline Formula/hledger.rb:39]: Do these resources just follow `hledger`'s version? If so, then I think you can also do `s/1.20.1/#{version}` here too.

**Review by carlocab — COMMENTED** [review]
Glad to see it's finally building. Just a few minor suggestions.

File: Formula/hledger.rb (hunk 2 above)
Comment (simonmichael) [reply Formula/hledger.rb]: Thanks, done.

File: Formula/hledger.rb (hunk 3 above)
Comment (simonmichael) [reply Formula/hledger.rb]: Done.

File: Formula/hledger.rb (hunk 4 above)
Comment (simonmichael) [reply Formula/hledger.rb]: Done.

File: Formula/hledger.rb (hunk 5 above)
Comment (simonmichael) [reply Formula/hledger.rb:39]: Not always, sometimes individual packages have their own minor version, so I think leaving them explicit is right.

File: Formula/hledger.rb (hunk 2 above)
Comment (carlocab) [reply Formula/hledger.rb]: Thanks. Sorry, one last thing, I think my suggestion may have obscured this in my previous comment:

>Could you leave a brief comment before this that explains how these values are supposed to be updated?

Comment (carlocab) [reply Formula/hledger.rb]: Excellent; thank you!

**Review by carlocab — DISMISSED** [review]
Thanks for your contribution, @simonmichael!

**Comment by simonmichael** [comment]
Thank you!

**Pull request closed without merging**